include/seqan3/alignment/multiple/library.hpp

```diff
@@ -0,0 +1,427 @@
+        {
+            return !(lhs < rhs);
+        }
+        //!\}
+    }; // class msa_library_iterator
+
+public:
+    //!\brief The bidirectional iterator that allows to visit all entries of the library.
+    using iterator = msa_library_iterator;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr msa_library()                                      = default; //!< Defaulted.
+    constexpr msa_library(msa_library const & other)             = default; //!< Defaulted.
+    constexpr msa_library(msa_library &&) noexcept               = default; //!< Defaulted.
+    constexpr msa_library & operator=(msa_library const & other) = default; //!< Defaulted.
+    constexpr msa_library & operator=(msa_library &&) noexcept   = default; //!< Defaulted.
+    ~msa_library()                                               = default; //!< Defaulted.
+    //!\}
+
+    /*!\brief Insert a new entry into the library.
+     * \param seq A pair of sequence indeices.
+     * \param pos A pair of position indices.
+     * \param score The assigned score for the specified positions in the specified sequences.
+     * \return True if the insertion took place, False if a score has already been assigned.
+     */
+    bool insert(coord_type seq, coord_type pos, score_type score)
+    {
+        check_seq_order(seq, pos);
+
+        // try to insert a new sequence pair
+        auto [iter, inserted] = data.try_emplace(seq, map_pos_score_type{{pos, score}});
+
+        if (inserted)    // we have added a new sequence pair with the entry
+            return true;
+        else             // append the score to the existing sequences and return whether it was successful
+            return iter->second.try_emplace(pos, score).second;
+    }
+
+    /*!\brief Add a score to a possibly existing entry in the library.
+     * \param seq A pair of sequence indeices.
+     * \param pos A pair of position indices.
+     * \param score The score that is added to the specified position in the specified sequence pair.
+     *              If this location does not exist it will be created with value `score`.
+     */
+    void add(coord_type seq, coord_type pos, score_type score)
+    {
+        check_seq_order(seq, pos);
+
+        // try to insert a new sequence pair
+        auto [iter, inserted] = data.try_emplace(seq, map_pos_score_type{{pos, score}});
+
+        if (!inserted)   // append to an existing sequence pair
+            iter->second[pos] += score;
+    }
+
+    /*!
+     * \brief Retrieve the iterator that contains the score of pairing two positions in the specified sequences.
+     * \param seq_pos A quadruple consisting of (seq1, seq2, pos1, pos2).
+     * \return An iterator to the score if the specified entry exists and end() otherwise.
+     */
+    iterator operator[](std::tuple<size_t, size_t, size_t, size_t> seq_pos) noexcept
```

Comment:
I think an  `operator[]` should always return a reference to a value (not an iterator)?